### PR TITLE
Add dnscrypt-proxy2 package.

### PR DESCRIPTION
**Pull request by waynr**
I am adding a new package instead of simply bumping the version of the old one for two reasons:

* The 1.x series of dnscrypt-proxy is written in C whereas the 2.x series is written in golang
* 2.x has different configuration conventions than 1.x; replacing the existing dnscrypt-proxy package with a backwards incompatible version would eventually break someone's preferred encrypted DNS setup.

I was reading some blog posts about encrypted DNS which recommended the 2.x series of dnscrypt rather than the 1.x. I also wanted an excuse to make my first contribution to this repository.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by waynr** [comment]
@joachifm @jgeerds since you two are maintainers on dnscrypt-proxy 1.x, I thought you might want to review this.

**Review by pSub — COMMENTED** [review]
Looks good to me.

**Comment by Mic92** [comment]
Should we keep the old version at all, it does not seem that it is maintained.

**Comment by Mic92** [comment]
@GrahamcOfBorg build dnscrypt-proxy2

**Comment by waynr** [comment]
@Mic92 I have no opinion on that; I didn't remove it because I didn't want to disrupt anyone who might be using it currently and it doesn't seem like much effort to just add a second package. @joachifm or @jgeerds might have an opinion on the matter.

File: pkgs/tools/networking/dnscrypt-proxy/2.x/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{ stdenv, buildGoPackage, fetchFromGitHub, go }:
```
Comment (worldofpeace) [inline pkgs/tools/networking/dnscrypt-proxy/2.x/default.nix]: I don't think you need go here

**Comment by worldofpeace** [comment]
Isn't it standard to use a ``` goDeps = ./deps.nix;``` in go packages? Especially one like this.

**Comment by worldofpeace** [comment]
@Mic92 There's a [nixos module](https://github.com/NixOS/nixpkgs/blob/2d3db84ddbb3555e48464b5cc20747f71c458c28/nixos/modules/services/networking/dnscrypt-proxy.nix) and [tests](https://github.com/NixOS/nixpkgs/blob/fef6b9ac0c58e59405023d866dfcccfa4b6fe9a6/nixos/tests/dnscrypt-proxy.nix)
for the original, so I don't think anyone wants that removed.

**Comment by waynr** [comment]
> Isn't it standard to use a goDeps = ./deps.nix; in go packages? 

Do you have a link to what the expected contents of `./deps.nix` would be? Based on the name, it sounds like the expectation would be that we should specify the dependencies of dnscrypt-proxy in this file which sounds wrong from the perspective of someone who spends every day writing, debugging, or otherwise interacting with go source code. Coming recently from a non-go background I know it sounds stupid/absurd but the modern convention of golang dependency management (even with `dep`) is for projects to vendor their dependencies in-repo; case in point, here is where dnscrypt-proxy keeps its dependencies: https://github.com/jedisct1/dnscrypt-proxy/tree/master/vendor

Also, for what it's worth I didn't see anything like this in the package whose example I was largely mimicking for this PR: https://github.com/NixOS/nixpkgs/tree/master/pkgs/applications/version-management/git-and-tools/hub

File: pkgs/tools/networking/dnscrypt-proxy/2.x/default.nix (hunk 1 above)
Comment (waynr) [reply pkgs/tools/networking/dnscrypt-proxy/2.x/default.nix]: Okay, I'm trying to rebuild without it but removing it doesn't seem to invalidate the already-built binary.

File: pkgs/tools/networking/dnscrypt-proxy/2.x/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  goPackagePath = "github.com/jedisct1/dnscrypt-proxy";
+
+  src = fetchFromGitHub {
+    owner = "jedisct1";
+    repo = "dnscrypt-proxy";
+    rev = "${version}";
+    sha256 = "0iwvndk1h550zmwhwablb0smv9n2l51hqbmzj354mcgi6frnx819";
+  };
+
+  enableParallelBuilding = true;
```
Comment (worldofpeace) [inline pkgs/tools/networking/dnscrypt-proxy/2.x/default.nix]: no need to set this as it's default https://github.com/NixOS/nixpkgs/blob/92609ec9ae725da937cdca7a4dcca1129a8bb919/pkgs/development/go-modules/generic/default.nix#L7

**Review by worldofpeace — APPROVED** [review]

**Comment by joachifm** [comment]
Thank you

**Comment by joachifm** [comment]
See also https://github.com/NixOS/nixpkgs/issues/43298

**Pull request merged**